File: Linked List/Remove_Nth_Last_Node.py
# Remove nth node from end of list
# EXAMPLE - remove 3RD node from end of list
# [1] -> [2] -> [3] -> [4] -> [5] -> [6] -> [7] -> [8] -> [9] -> [10] -> NONE
#                                                   ^ remove so that: ...[7] -> [9] ->...
# Step 1) create a dummy node to act as head. Create 2 pointers - left & right.
#           [dummy] -> [1] -> [2] -> ...
#              ^        ^
#             left     right
#
# Step 2) while n > 0, traverse right variable. Each traversal, n -= 1.
#           Once loop ends, right variable will be on [4].
# Step 3) then while right != None, traverse right & left, like below:
#           [1] -> [2] -> [3] -> [4] -> [5] -> ...
#                >>left>>             >>right>>
#                          |_____________| <- this is the size of n
#
# Step 4) once right is None, then have left.next = left.next.next
#
#           ... [7] -> [8] -> [9] -> [10] -> NONE
#               left                         right
#
#
#           ... [7] --------> [9] -> [10] -> NONE
#               left                         right
#
# Step 5) finally, return dummy.next (so the first empty head node is ignored).

class Node:
    def __init__(self, value=None):
        self.value = value
        self.next = None

    def __repr__(self):
        return f"NODE: {self.value}"


class LinkedList:
    def __init__(self):
        self.head = None

    def print_list(self):
        if self.head:
            current = self.head
            while current:
                print(current.value, end=" -> ")
                current = current.next
            print("NONE")

    def append(self, value):
        new_node = Node(value)
        if self.head:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        else:
            self.head = new_node

    def remove_nth_from_end(self, n):
        dummy = Node().next = self.head
        left = dummy
        right = dummy.next

        while right and n > 0:
            right = right.next
            n -= 1

        while right:
            right = right.next
            left = left.next

        left.next = left.next.next

        return dummy


def print_list(head):
    if head:
        current = head
        while current:
            print(current.value, end=" -> ")
            current = current.next
        print("NONE")


node = LinkedList()
# create nodes
for i in range(1, 11):
    node.append(i)


dum = node.remove_nth_from_end(1)
print_list(dum)

File: C Data Structures _ Algorithms/removeNthLastNode.c
#include <stdlib.h>
#include <stdio.h>

typedef struct Node{
    int value;
    struct Node* next;
}Node;
void printL(Node* head){
    Node* current = NULL;
    current = head;
    printf("LINKED LIST: ");
    while (current){
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL\n");
}
void append(Node** head, int value){
    // Create Node & fill it
    Node* node = NULL;
    node = (Node* ) malloc(sizeof(Node));
    node->next = NULL;
    node->value = value;

    Node* curr = NULL;
    if (*head){
        curr = *head;
        while (curr->next){
            curr = curr->next;
        }
        curr->next = node;
    }
    else {
        *head = node;
    }
}
void removeNthLastNode(Node** dummy, Node* head, int n){
    *dummy = (Node* ) malloc(sizeof(Node));
    (*dummy)->next = head;
    Node* left = *dummy;
    Node* right = (*dummy)->next;
    while (right && n > 0){
        right = right->next;
        n--;
    }
    while (right){
        left = left->next;
        right = right->next;
    }
    left->next = left->next->next;
}

int main(){
    Node* head = NULL;
    // Linked List:
    append(&head, 1); append(&head, 2); append(&head, 3); append(&head, 4); append(&head, 5); append(&head, 6); append(&head, 7); append(&head, 8); append(&head, 9); append(&head, 10);
    // remove nth last node
    Node* dummy = NULL;
    removeNthLastNode(&dummy, head, 3);
    printL(dummy->next);
    return 0;
}